### PR TITLE
CAMEL-20904: Add Kubernetes Camel JBang plugin

### PR DESCRIPTION
**Pull request by christophd**
# Description

This PR introduces a new Camel JBang plugin to deploy/run on Kubernetes. The plugin extends the Camel JBang export functionality to also generate a Kubernetes manifest (kubernetes.yml) which holds all resources (Deployment, Service, ConfigMap, etc.) to run the application on Kubernetes.

The idea is to run a project export first and then build and deploy via SB / Quarkus standards using the generated Kubernetes manifest just like how you would normally do this via their Maven plugins.

The Kubernetes plugin provides the following commands:
- export to generate the Quarkus/SB project with the Kubernetes manifest being part of the exported project
- run to export the project in the background and call Maven package to build and deploy to Kubernetes (also supports `--dev` option to use automatic redeployment)
- delete to remove all resources from Kubernetes using the generated Kubernetes manifest
- logs to show the logs of the running Pod

Traits:
The Kubernetes export functionality is able to make use of the Camel K traits in order to simplify the Kubernetes manifest generation (e.g. adding envvars, configure exported ports, use volume mounts, secrets, configmaps etc.). At the moment the PR supports some basic traits (DeploymentTrait, ContainerTrait, EnvTrait, ...) more trait implementations may follow.

Detailed changes in this PR:
- Introduce new Camel JBang plugin `camel-jbang-plugin-kubernetes` to configure/run Camel applications on Kubernetes
- Camel K JBang plugin depends on the new Kubernetes plugin
- Move Kubernetes basics (e.g. PodLogs command) from Camel K plugin to the new Kubernetes plugin
- Kubernetes plugin provides run, delete, export and logs command
- Kubernetes export command extends the arbitrary project export command and adds the generation of a Kubernetes manifest (kubernetes.yml) that becomes part of the exported project.
- Kubernetes run command uses the exported project to build the application image (with container-jib)
- Kubernetes deployment is done with arbitrary Quarkus/Spring Boot tooling and uses the generated Kubernetes manifest (kubernetes.yml)
- Kubernetes delete command uses the manifest (kubernetes.yml) to remove all resources
- Kubernetes logs command connects to the pod (selected via label conventions) and displays the log output

The PR put focus on Maven based projects with Quarkus platform as a runtime. Next step would be to also apply this Kubernetes run mechanism to Spring Boot (eventually camel-main).

TODOs in this PR:
- remove Camel K CRD model dependency in favor of using own trait domain model in Camel JBang
- make image builder configurable to support builders other than quarkus-container-image-jib
- remove fixed kubernetes-client version as soon as 6.13.1 is part of the Quarkus platform
- implement SpringBoot project build & Kubernetes deployment

This is considered to be a draft PR in order to find agreement within the team and to do some tuning with adding documentation and writing some unit tests.

# Target

- [x] I checked that the commit is targeting the correct branch (note that Camel 3 uses `camel-3.x`, whereas Camel 4 uses the `main` branch)

# Tracking
- [x] If this is a large change, bug fix, or code improvement, I checked there is a [JIRA issue](https://issues.apache.org/jira/browse/CAMEL) filed for the change (usually before you start working on it).

See CAMEL-20904

# Apache Camel coding standards and style

- [x] I checked that each commit in the pull request has a meaningful subject line and body.

- [x] I have run `mvn clean install -DskipTests` locally and I have committed all auto-generated changes

### CODE REVIEW

**Review by oscerd — APPROVED** [review]
Looks really promising. Thanks @christophd

File: dsl/camel-jbang/camel-jbang-plugin-k/pom.xml [hunk 1]
```diff
@@ -46,6 +46,13 @@
           <artifactId>camel-jbang-core</artifactId>
         </dependency>
 
+        <!-- TODO: clarify why jbang plugin is not part of camel-bom -->
```
Comment (davsclaus) [inline dsl/camel-jbang/camel-jbang-plugin-k/pom.xml:49]: Add a JIRA so we can add it to camel-bam

File: dsl/camel-jbang/camel-jbang-plugin-kubernetes/pom.xml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>camel-jbang-plugin-kubernetes</artifactId>
+
+    <name>Camel :: JBang :: Plugin :: Kubernetes</name>
+    <description>Camel JBang Kubernetes Plugin</description>
+
+    <properties>
+        <firstVersion>4.4.0</firstVersion>
```
Comment (davsclaus) [inline dsl/camel-jbang/camel-jbang-plugin-kubernetes/pom.xml:37]: 4.8.0

File: dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesDelete.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import io.fabric8.kubernetes.api.model.StatusDetails;
+import org.apache.camel.dsl.jbang.core.commands.CamelJBangMain;
+import org.apache.camel.util.FileUtil;
+import org.apache.camel.util.StringHelper;
+import picocli.CommandLine;
+
+@CommandLine.Command(name = "delete", description = "Delete Camel application from Kubernetes", sortOptions = false)
+public class KubernetesDelete extends KubernetesBaseCommand {
+
+    @CommandLine.Parameters(description = "The Camel file(s) to run.",
```
Comment (davsclaus) [inline dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesDelete.java:33]: delete

File: dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesExport.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+            return StringHelper.afterLast(image, ":");
+        }
+
+        return super.getVersion();
+    }
+
+    /**
+     * Configurer used to customize internal options for the Export command.
+     *
+     * @param runtime
```
Comment (davsclaus) [inline dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesExport.java:255]: All this param javadoc is not really needed

File: dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesRun.java [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    String[] filePaths;
+
+    @CommandLine.Option(names = { "--trait-profile" }, description = "The trait profile to use for the deployment.")
+    String traitProfile;
+
+    @CommandLine.Option(names = { "--dependency", "-d" },
+                        description = "Adds dependency that should be included, use \"camel:\" prefix for a Camel component, \"mvn:org.my:app:1.0\" for a Maven dependency.")
+    String[] dependencies;
+
+    @CommandLine.Option(names = { "--property", "-p" },
```
Comment (davsclaus) [inline dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesRun.java:56]: We do not favour single char options, we use `--prop` and `--dep` / `--deps`
See the vanilla run command

File: dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesRun.java [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+            return exit;
+        }
+
+        if (dev) {
+            DefaultCamelContext reloadContext = new DefaultCamelContext(false);
+            configureFileWatch(reloadContext, export, workingDir);
+            reloadContext.start();
+        }
+
+        if (dev || wait || logs) {
```
Comment (davsclaus) [inline dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesRun.java:237]: If you run in `--dev` mode, does the pod autoatic delete when you exit via ctrl + c ?

**Comment by davsclaus** [comment]
Great work @christophd - love that we can make it gitops compliant, and that the k8s experience becomes just a standard Q or SB maven project with that kubernetes yaml file. Must less that goes wrong, and its all done in the same way that users already do, or would do if they did it by hand.

File: dsl/camel-jbang/camel-jbang-core/src/main/java/org/apache/camel/dsl/jbang/core/commands/Export.java [hunk 7]
```diff
@@ -47,9 +49,9 @@ protected Integer export() throws Exception {
             System.err.println("The runtime option must be specified");
             return 1;
         }
+
         if (gav == null) {
-            System.err.println("The gav option must be specified");
-            return 1;
+            gav = "org.apache.camel:%s:%s".formatted(getProjectName(), getVersion());
```
[the remark this answers is not in the document]
Comment (christophd) [reply dsl/camel-jbang/camel-jbang-core/src/main/java/org/apache/camel/dsl/jbang/core/commands/Export.java:54]: > And then required = true should be removed from gav option so its optional.

👍 I think the option is optional already

[the remark this answers is not in the document]
Comment (christophd) [reply dsl/camel-jbang/camel-jbang-core/src/main/java/org/apache/camel/dsl/jbang/core/commands/Export.java:54]: What about `org.example.project` as a default? `org.apache.camel` is being used by SBOMGenerator command and in the automatic dummy exports that is part of the run command. Should we change these occurrences, too?

File: dsl/camel-jbang/camel-jbang-plugin-k/pom.xml (hunk 1 above)
Comment (christophd) [reply dsl/camel-jbang/camel-jbang-plugin-k/pom.xml:49]: 👍 https://issues.apache.org/jira/browse/CAMEL-20955

File: dsl/camel-jbang/camel-jbang-plugin-kubernetes/pom.xml (hunk 2 above)
Comment (christophd) [reply dsl/camel-jbang/camel-jbang-plugin-kubernetes/pom.xml:37]: Good catch!

File: dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesRun.java (hunk 6 above)
Comment (christophd) [reply dsl/camel-jbang/camel-jbang-plugin-kubernetes/src/main/java/org/apache/camel/dsl/jbang/core/commands/kubernetes/KubernetesRun.java:237]: No, it does not. This is a good idea. Let me see if I can add this, thx

File: dsl/camel-jbang/camel-jbang-core/src/main/java/org/apache/camel/dsl/jbang/core/commands/Export.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (davsclaus) [reply dsl/camel-jbang/camel-jbang-core/src/main/java/org/apache/camel/dsl/jbang/core/commands/Export.java:54]: Yeah good idea to change them all

**Comment by christophd** [comment]
Many thanks for the initial feedback. I think this one is ready for final review.

I have fixed some reported issues, added unit tests, added documentation and the ability to auto remove the deployment when terminating a process with `--dev` option.

**Review by davsclaus — APPROVED** [review]

**Pull request merged**